free/src/main/scala/cats/free/Free.scala

```diff
@@ -250,3 +264,57 @@ object Free {
       def flatMap[A, B](a: Free[S, A])(f: A => Free[S, B]): Free[S, B] = a.flatMap(f)
     }
 }
+
+private trait FreeFoldable[F[_]] extends Foldable[Free[F, ?]] {
+
+  implicit def F: Foldable[F]
+
+  override final def foldLeft[A, B](fa: Free[F, A], b: B)(f: (B, A) => B): B =
+    fa.foldStep(
+      a => f(b, a),
+      fa => F.foldLeft(fa, b)(f),
+      { case (fx, g) => F.foldLeft(fx, b)((bb, x) => foldLeft(g(x), bb)(f)) }
+    )
+
+  override final def foldRight[A, B](fa: Free[F, A], lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B] =
+    fa.foldStep(
+      a => f(a, lb),
+      fa => F.foldRight(fa, lb)(f),
+      { case (fx, g) => F.foldRight(fx, lb)( (a, lbb) => foldRight(g(a), lbb)(f)) }
```

Comment:
I think that `lbb.flatMap(bb => foldRight(g(a), Eval.now(bb))(f))` instead of `foldRight(g(a), lbb)(f)` should make this stack safe.

We should probably add a test to check this. Something like (haven't compiled this):

```scala
val n = 50000
val freeOption: Int => Free[Option, Int] = Free.pure(x)
val free = (0 to n).foldLeft(freeOption(0))((r, _) => r.flatMap(n => freeOption(n + 1)))
free.foldRight(0)((a, lb) => lb.map(_ + a)) should (n)
```